clean command

g-w1 opened the pull request.

marler8997 commented:
FYI, let's keep whitespace changes and functionality changes separate

marler8997 commented on zigup.zig:
```diff
@@ -200,8 +201,7 @@ pub fn main2() !u8 {
         return 1;
-    }
-    if (std.mem.eql(u8, "fetch-index", args[0])) {
+    } else if (std.mem.eql(u8, "fetch-index", args[0])) {
```
The elses are unnecessary here, since each of these code paths return from the function.

marler8997 commented on zigup.zig:
```diff
@@ -421,19 +420,48 @@ fn listCompilers(allocator: *Allocator) !void {
 }
 
-fn printDefaultCompiler(allocator: *Allocator) !void {
+fn cleanCompilers(allocator: *Allocator, version_name: [:0]u8) !void {
```
So, I'm not sure how useful it is to implementing cleaning a specific compiler version.  The user can do that easy enough by just deleting the directory themselves.  What's more useful is the ability to all unused compilers.  So this goes hand-in-hand with implementing the "keep" functionality.  My idea was to include an empty file, probably named "keep" in each compiler install directory if the user wants to keep it.  So the clean command would just clean all compilers that don't have a "keep" marker file.

g-w1 replied:
Ok. I see now. Ill try to impliment your idea with keep. If a user wants to keep a certain master, should they just give it with the hash, like `zigup keep 0.6.0+123456`? Thank you for all this feedback, I am becoming better. I have also removed the unnessary elses.

marler8997 replied:
>  If a user wants to keep a certain master, should they just give it with the hash, like zigup keep 0.6.0+123456?

Basically yeah, if they want to keep any version it's just `zigup keep VERSION`.  Maybe it would also work with `master`?  Not sure on that one, since the user may not have master downloaded....

marler8997 commented on zigup.zig:
```diff
@@ -222,6 +225,15 @@ pub fn main2() !u8 {
         return 0;
     }
+    if (std.mem.eql(u8, "clean", args[0])) {
+        if (args.len == 2) {
```
Code usually ends up being simpler if you check for errors first, like this:
```zig
        if (args.len != 2) {
            std.debug.warn("Error: 'clean' command requires 1 argument but got {}\n", .{args.len - 1});
            return 1;
        }
	try cleanCompilers(allocator, args[1]);
        return 0;
```

marler8997 commented on zigup.zig:
```diff
@@ -421,19 +428,78 @@ fn listCompilers(allocator: *Allocator) !void {
+    // getting/reading the keep file
+    const install_dir_string = try getInstallDir(allocator, .{ .create = true });
+    defer allocator.free(install_dir_string);
+    const keep_file_path = try std.fs.path.join(allocator, &[_][]const u8{ install_dir_string, "keep" });
```
The idea was that each compiler would have the keep file inside it's own install directory (not having a keep file with a list of version).

g-w1 replied:
Ah, I misread. Well this seems to be much easier to implement. Thanks

marler8997 commented:
Also be sure to utilize the "test" script to test all new features you add.

g-w1 commented:
I just added the test and fixed the script to look in the compiler directory for `keep`. The only thing that I am not sure of is how the master semantics work; when zigup finds 0.6.0+123456 and master directory, it doesn't know they are the same thing, so if one is default it deletes the one of them, deleting both. I could hardcode that it skips over master when looking for things to delete, do you think this would be a good idea?

marler8997 commented on zigup.zig:
```diff
@@ -421,19 +425,75 @@ fn listCompilers(allocator: *Allocator) !void {
+            }
+        }
+        // if we didn't continue at all then actually delete the compiler
+        const abs_path_to_delete = try std.fs.path.join(allocator, &[_][]const u8{ install_dir_string, entry.name });
```
I think it would be better if we created this path above (before checking for keep file), then we created the `look_for_keep_path` based on this path.

marler8997 commented on zigup.zig:
```diff
@@ -421,19 +425,75 @@ fn listCompilers(allocator: *Allocator) !void {
-    var targetPathBuffer: [std.fs.MAX_PATH_BYTES]u8 = undefined;
-    if (std.fs.readLinkAbsolute(pathLink, &targetPathBuffer)) |targetPath| {
-        std.debug.warn("{}", .{std.fs.path.basename(std.fs.path.dirname(std.fs.path.dirname(targetPath).?).?)});
+    // var targetPathBuffer: [std.fs.MAX_PATH_BYTES]u8 = undefined;
```
Why change this?

g-w1 replied:
It returns a stack allocated slice, so I had to manually allocate it. I am going to change this to inline fn so that it can do this.

g-w1 replied:
I changed it to take a buffer after chatting with people on irc about best way to do this.

marler8997 replied:
gotcha I see now

marler8997 commented on zigup.zig:
```diff
@@ -241,17 +250,12 @@ pub fn main2() !u8 {
-            } else {
-                try setDefaultCompiler(allocator, compilerDir);
-            }
+            try setDefaultCompiler(allocator, compilerDir);
```
I don't think this will work on windows....because of that, I'm not sure we want to support setting the default to "master", or at least, come up with a solution that will work without requiring symlinks like this.

Your change should still work without this particular change, we can create separate Pull Requests for them.  In general, better to keep each pull request focused on one thing.  For this PR, the only change I expect is implementing the clean command itself.

g-w1 replied:
Ok I understand. Thanks

marler8997 commented on zigup.zig:
```diff
@@ -246,12 +255,12 @@ pub fn main2() !u8 {
             } else {
                 try setDefaultCompiler(allocator, compilerDir);
             }
+            try setDefaultCompiler(allocator, compilerDir);
```
You left this line....now you're setting the default compiler twice

marler8997 reviewed (approved):
thanks for the work you put into this
